### DISCUSSION TITLE
General questions about troubleshooting

### DISCUSSION THREAD

**Discussion by andrefmferreira** [igorocampos/ESCPOS#33] [General]
Hi @igorocampos ! Great project. many thanks for that.
i've been trying to solve an issue regarding escpos qrcode printing, because for some reason i don't know how, i can't print more than 124 chars of data in a qrcode.

i've installed you nuget, and the result is the same. Do you know how, using your library, can i print more than 124 chars?

Regards

**Comment by igorocampos** [comment]
@andrefmferreira 
Which QRCode type are you using? Micro, 1 or 2?

QRCode Model2 would be the proper one to be using with such a long string.

**Reply by andrefmferreira** [reply]
I've tested both with 1 and 2. 

regards

**Reply by igorocampos** [reply]
@andrefmferreira 
What printer maker and model are you on? What happens when you use 125+ chars in qrcode data? Does it get something printed or just nothing? Have you tried changing the correction percentage and size parameters?

I don't own a printer anymore, so I can't test it, but it should be working, I know a few people that are using successfully this library in production, especially for SAT CF-e receipts, which require more than 124 chars of data. 
Lastly, I would say to check the bytes right before they are sent to the printer, do they match what is expected according to the ESC/POS [manual](https://github.com/igorocampos/ESCPOS/blob/master/ESCPOS_Command_Manual.pdf)? If yes, perhaps it could be a hardware issue.

**Reply by andrefmferreira** [reply]
@igorocampos when i print 125+ chars, it prints the qrcode, along with text before it (the text that exceeds the 124 chars in the string). i've used for now 2 printers. a a Sumni v2 pro and a HPRT HM-E300

Regarding the bytes, i think the issue is regarding the data.length % 256. its value is 135 (132 +3). the way i see it, its truncating on 127. 

i've mount the qr functions, and it works as your library does.

Just for notice, i use a printing service (the software's OS is android. in this case, we use a service called RAWBT. We print the data into a text file, and the service just prints it. 
If i print directly to the COM port its the same

Many thanks for your time ;)

Regards

**Reply by igorocampos** [reply]
Alright, so let's go step by step:
1 - You are using 132 chars for QRCode data, right?
2 - Code adds 3 to that number, so `num` = 135
3 - `pL` = num % 256, so pL will be `135` (the remainder of 135 divided by 256)
4 - `pH` = num / 256, so pH will be `0`
5 - `storeData` = { 0x1D, 0x28, 0x6B, 0x87, 0x00, 0x31, 0x50, 0x30 };

I don't see any truncation so far. Are you able to debug if `storeData` variable contains the above bytes?

The sum matches the manual requirements as well
![image](https://user-images.githubusercontent.com/10572656/207588699-87e8f1c6-b41f-4e85-b42c-91c8e15a0e2d.png)


About printing it on Android, that's fine, I also have successfully printed it on a Sunmi v2, however I didn't use any 3rd party service for printing, I just sent the bytes to the inner printer via Bluetooth. I can share that code with you, if you want to try it that way.

My guess is that service of yours (RAWBT) is the one that might not be retrieving all bytes from a text file correctly. Perhaps it has an issue with `0x00` ? I've seen this in text files in the past, when they have a null byte (0x00) the file malfunctions.

@andrefmferreira

**Reply by igorocampos** [reply]
@andrefmferreira 

Here's the mentioned code to print in the Sunmi v2 

```cs
using System;
using Android.Bluetooth;
using Java.Util;

public static class BluetoothUtil
{
    private static UUID PRINTER_UUID = UUID.FromString("00001101-0000-1000-8000-00805F9B34FB");
    
    private static BluetoothDevice GetDevice(BluetoothAdapter bluetoothAdapter)
    {
        foreach (BluetoothDevice device in bluetoothAdapter.BondedDevices)
        {
            if (device.Address == Android.OS.Build.Model.StartsWith("APOS") ? "00:01:02:03:0A:0B" : "00:11:22:33:44:55")
                return device;
        }
        return null;
    }

    private static BluetoothSocket GetSocket(BluetoothDevice device)
    {
        BluetoothSocket socket = device.CreateRfcommSocketToServiceRecord(PRINTER_UUID);
        socket.Connect();
        return socket;
    }

    private static void SendData(byte[] bytes, BluetoothSocket socket)
    {
        var o = socket.OutputStream;
        o.Write(bytes, 0, bytes.Length);
        o.Close();
    }

    public static bool Imprimir(byte[] impressao, string titulo = "")
    {
        BluetoothAdapter btAdapter = BluetoothAdapter.DefaultAdapter;

        if (btAdapter == null)
        {
            // "Não foi possivel imprimir {titulo}: Bluetooth desligado!"
            return false;
        }

        BluetoothDevice device = BluetoothUtil.GetDevice(btAdapter);
        if (device == null)
        {
            // "Não foi possivel imprimir {titulo}: Impressora não encontrada!"
            return false;
        }

        BluetoothSocket socket = BluetoothUtil.GetSocket(device);
        SendData(impressao, socket);
        return true;
    }
}
```

To use it, you can just do:

```cs
byte[] escPosCommand = new byte[]{}; //Replace this with the bytes you wanna send to the printer
BluetoothUtil.Imprimir(escPosCommand);
```

**Reply by andrefmferreira** [reply]
HI @igorocampos ,

Sorry for the late answer. It were busy days.

I've tested your code and it works perfectly. The way i see it, my problem is either with the service "RAW BT", or the BinaryWriter, wich writes the bytes into a text file (to rawbt reads).

In your experience do you have any knoladge about RawBT?

Anyway, since your code runs perfectly, i wont abset you further. I'will try to send data directly to the device instead of using a 3rd party service

Many many thanks!!

Regards and have your self a very merry christmas!
Feliz Natal ;)

Regards

**Reply by igorocampos** [reply]
Hi @andrefmferreira, I'm glad to hear that it worked! Unfortunately, I don't have any experience using RAW BT service... But the first thing I would check is the buffer size, perhaps in the service configuration. If the buffer is only 124 bytes, that explains the behavior. About BinaryWritter I believe it's straightforward, but if you want to double-check it, just copy the file to your PC and see if all bytes are in the file.

Feliz Natal, 

Abraços,
Igor

But yes, my advice would be to keep sending yourself the bytes instead of depending on things you don't own/have 100% control over.

**Comment by andrefmferreira** [comment]
HI @igorocampos !
i'have a question about your project. if i use the PrintQRCode method, it prints a small qrcode
if i add parameters such as PrintQRCode("string", QRCodeModel.Model1, qrCodeSize: QRCodeSize.Large) it prints a bigger one, but yet with 2.3cm HxW.
So if i want to print it bigger, what do you recommend?
Also, is it possible, using your library to print to a usb device in android?

Best regards!

Obrigado ;)
André

**Reply by igorocampos** [reply]
Hi @andrefmferreira, 
I believe this is the command you want to change:

![image](https://github.com/igorocampos/ESCPOS/assets/10572656/0143185c-c85f-473c-9918-396020f9afb7)


Currently, the `n` parameter is dictated by the enum you mentioned, QRCodeSize which ranges from 2 to 5.

![image](https://github.com/igorocampos/ESCPOS/assets/10572656/1a4e102f-f95a-4ddd-bbc5-1b4b5f168726)

 If you wish you can create a PR adding new values for that enum. The manual doesn't say anything about limits, but I would test it 1 by 1, until you have the desired size that works.
 
 Regarding the other question:
> using your library to print to a usb device in android?

The library just send bytes to a destination, like a serial port, Bluetooth socket, a path in the network, or even a file. I never had the need to use USB ports as a destination, but in theory, it should be about the same, as long as it accepts bytes as an input.

**Reply by andrefmferreira** [reply]
Worked like a charm!! Thanks @igorocampos. the byte that works best (~32mm) is 7. 
Regards!!
Obrigado ;)

**Comment by andrefmferreira** [comment]
Hi @igorocampos, hope to dont bother you to much.

As i understand, your library uses UTF8, but i'm having trouble printing special chars as '@' or other portuguese chars like in this word: 'talão'

Any thougths?

Many thanks in advance!

**Reply by igorocampos** [reply]
I've never really run into that problem, however from other's issues I would say it's your Code Page. There's a command that you choose what code page to use in your printer. Have you given it a try? Or do you know what Code page are you using?

**Reply by igorocampos** [reply]
hey @andrefmferreira I totally missed this last message of yours... Sorry, but in the latest version I've implemented a better way to deal with different encodings, there's some explanations in the README file teaching how to do it.
